### ISSUE TITLE
Varnish & apache + api-platform invalidation: Cache-Tags too big

### ISSUE THREAD

**Issue by jocel1** [api-platform/api-platform#1046]
Hi!

I'm trying to use api-platform with varnish.
I'm using the bundled default.vcl file and the following php config :
```
    http_cache:
      invalidation:
        enabled: true
        varnish_urls: ['%env(VARNISH_URL)%']
      # Adds sensitive default cache headers
      max_age: 0
      shared_max_age: 3600
      vary: ['Content-Type', 'Authorization']
      public: true
```

However on specific api call, I'm getting a 500 error each time the invalidation is enabled.
On apache, it just says:
```
[Wed Mar 13 22:49:24.402613 2019] [proxy_fcgi:error] [pid 8532:tid 139994769344256] [client 127.0.0.1:47138] Premature end of script headers: index.php
[Wed Mar 13 22:49:24.402645 2019] [proxy_fcgi:error] [pid 8532:tid 139994769344256] [client 127.0.0.1:47138] AH01070: Error parsing script headers
[Wed Mar 13 22:49:24.402649 2019] [proxy_fcgi:error] [pid 8532:tid 139994769344256] (22)Invalid argument: [client 127.0.0.1:47138] AH01075: Error dispatching request to : ```

And on Varnish while requesting this url:
```
*   << Request  >> 32794     
-   Begin          req 32793 rxreq
-   Timestamp      Start: 1552517731.654205 0.000000 0.000000
-   Timestamp      Req: 1552517731.654205 0.000000 0.000000
-   ReqStart       172.31.47.11 51372 a0
-   ReqMethod      GET
-   ReqURL         /api/mums/zRGKqZvGFGdgd2z8pJmcF1QKGCS2
-   ReqProtocol    HTTP/1.1
-   ReqHeader      X-Forwarded-For: 37.169.147.162, 108.162.229.246
-   ReqHeader      X-Forwarded-Proto: https
-   ReqHeader      X-Forwarded-Port: 443
-   ReqHeader      Host: myhost.com
-   ReqHeader      X-Amzn-Trace-Id: Root=1-5c898a63-f19190d7b3c39f56e4c730d0
-   ReqHeader      Accept-Encoding: gzip
-   ReqHeader      CF-IPCountry: FR
-   ReqHeader      CF-RAY: 4b71988eb8b53c2f-CDG
-   ReqHeader      CF-Visitor: {"scheme":"https"}
-   ReqHeader      user-agent: curl/7.54.0
-   ReqHeader      accept: application/ld+json
-   ReqHeader      x-auth-token: 4UFsaVeqleSIMbvRqBfnlZ0mnE
-   ReqHeader      CF-Connecting-IP: 37.169.147.162
-   ReqHeader      CDN-Loop: cloudflare
-   ReqUnset       X-Forwarded-For: 37.169.147.162, 108.162.229.246
-   ReqHeader      X-Forwarded-For: 37.169.147.162, 108.162.229.246, 172.31.47.11
-   VCL_call       RECV
-   VCL_return     hash
-   VCL_call       HASH
-   VCL_return     lookup
-   HitMiss        32786 -247.251404
-   VCL_call       MISS
-   VCL_return     fetch
-   Link           bereq 32795 fetch
-   Timestamp      Fetch: 1552517731.777657 0.123452 0.123452
-   RespProtocol   HTTP/1.1
-   RespStatus     500
-   RespReason     Internal Server Error
-   RespHeader     Date: Wed, 13 Mar 2019 22:55:31 GMT
-   RespHeader     Server: Apache/2.4.25 (Debian)
-   RespHeader     ETag: "e885a24ca2e41972f78b8f84b5dca611"
-   RespHeader     Content-Length: 0
-   RespHeader     Content-Type: application/ld+json; charset=utf-8
-   RespHeader     url: /api/mums/zRGKqZvGFGdgd2z8pJmcF1QKGCS2
-   RespHeader     X-Varnish: 32794
-   RespHeader     Age: 0
-   RespHeader     Via: 1.1 varnish (Varnish/6.0)
-   VCL_call       DELIVER
-   RespUnset      url: /api/mums/zRGKqZvGFGdgd2z8pJmcF1QKGCS2
-   VCL_return     deliver
-   Timestamp      Process: 1552517731.777670 0.123465 0.000013
-   RespHeader     Connection: keep-alive
-   Timestamp      Resp: 1552517731.777684 0.123479 0.000014
-   ReqAcct        498 0 498 300 0 300
-   End    
```

It seems to be related to the very long cache-tags list which could be returned in this case because of multiple nested objects / lists. I guess there's no way to fix this, other than switching to nginx?

**Comment by gavinlove** [comment]
You can increase the max header size in varnish.

https://varnish-cache.org/docs/4.1/reference/varnishd.html#http-req-hdr-len
https://varnish-cache.org/docs/4.1/reference/varnishd.html#http-max-hdr

**Comment by teohhanhui** [comment]
@gavinlove He's hitting a limit in Apache, not Varnish.

**Comment by teohhanhui** [comment]
I think it might be this one:

Parameter | Default | Description
-- | -- | --
responsefieldsize | 8192 | Adjust the size of the proxy response field buffer. The buffer size             should be at least the size of the largest expected header size from             a proxied response. Setting the value to 0 will use the system             default of 8192 bytes.         Available in Apache HTTP Server 2.4.34 and later.

https://httpd.apache.org/docs/2.4/mod/mod_proxy.html

But I've no idea how it should be set when using `SetHandler` (like in the [Symfony docs](https://symfony.com/doc/current/setup/web_server_configuration.html#using-mod-proxy-fcgi-with-apache-2-4))

**Comment by jocel1** [comment]
@teohhanhui Thanks for the tips.
I've just found
```
<Proxy "fcgi://127.0.0.1:9000">
  ProxySet responsefieldsize=16384
</Proxy>

<FilesMatch \.php$>
  SetHandler proxy:fcgi://127.0.0.1:9000
</FilesMatch>``` 

I'll give it a try

**Comment by jocel1** [comment]
Just tested with

```
    <Proxy "unix:/run/php/php7.2-fpm.sock|fcgi://localhost">
        ProxySet responsefieldsize=16384
    </Proxy>

    <FilesMatch ".+\.ph(ar|p|tml)$">
        SetHandler "proxy:unix:/run/php/php7.2-fpm.sock|fcgi://localhost"
    </FilesMatch>
```
But it's not working

**Comment by cristiroma** [comment]
@jocel1 - have you tried with 2.4.34+ version?

**Comment by jocel1** [comment]
@cristiroma yes it was a 2.4.34+ version with proper support of the responsefieldsize parameter

**Comment by mrpatrick** [comment]
Having the same issue as @jocel1 and does't seem to be working for me either.

**Comment by teohhanhui** [comment]
Try nginx instead. I don't know much about Apache httpd and we obviously couldn't figure out the solution.

**Comment by dhpowrhost** [comment]
Anyone managed to get it working with apache?
Im on Apache 2.4.41 and the issue persists - even with responsefieldsize set much higher than default.

**Comment by keith4** [comment]
In case anyone else stumbles onto this: 'responsefieldsize' is only implemented for mod_proxy_http

**Comment by jocel1** [comment]
And BTW you can control the behavior directly in api-platform thanks to this PR : https://github.com/api-platform/core/pull/2865

**Comment by soyuka** [comment]
Closing per #2987 — this repository is becoming a CLI installer. The previous Docker/template distribution is being superseded by the installer + upstream [symfony-docker](https://github.com/dunglas/symfony-docker). If the issue still reproduces with the new setup, please open it on `dunglas/symfony-docker` (for Docker/Caddy/FrankenPHP issues) or `api-platform/core` (for framework issues).
